### PR TITLE
Change `range.size` to return `int`; add `range.sizeAs()`

### LINKED ISSUE #7530: range.size can overflow
### Summary of Problem
The length of maximal uint ranges such as `0:uint(8)..255:uint(8)` overflows.
This is currently because the return type of length is the index type of the
range. You might be able to return `chpl__maxIntTypeSameSign(idxType)` for the
smaller uints, but for the `uint(64)` case, there's really nothing you can do
about it though.

This overflow is more annoying for int ranges where things like
`(0:int(8)..127..int(8)).length` result in overflow. Currently this case isn't
being tested in the .future because integer overflow is undefined.
I'm not sure if there's actually anything we can do about this issue over
having range.length returning uint(64)/int(64) instead of the index type but
that only masks the problem for some ranges.

I think I'd at least prefer some error checking that will report that overflow
is going to occur so the user isn't completely surprised by this.

**Associated Future Test(s):**
 [`test/types/range/elliot/rangeLength/rangeLengthOverflow.chpl`](https://github.com/chapel-lang/chapel/blob/master/test/types/range/elliot/rangeLength/rangeLengthOverflow.chpl)

### LINKED ISSUE #17264: Can't create array over int(8) domain with more than 127 elements
### Summary of Problem

**Source Code:**

```chapel
var lo = -0x1:int(8);
var hi = 0x7e:int(8);
writeln((lo,hi));

var A : [lo..hi] int;
writeln(A);
```

```
(-1, 126)
$CHPL_HOME/modules/internal/ChapelBase.chpl:960: error: Attempting to allocate > max(size_t) bytes of memory
```

I added a `writeln(size)` to `ChapelBase.chpl`, which reports `-128`.

It works if hi is `0x7d`, reporting a size of `127`.

**Associated Future Test(s):**

None yet.

### Configuration Information

```
chpl version 1.24.0 pre-release (57a3573b44)

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

gcc (Ubuntu 10.2.0-13ubuntu1) 10.2.0
```

### PR DESCRIPTION
**Pull request by bradcray**
This PR changes `.size` routines on ranges, domains, and arrays to return `int` by default under the rationale that it's pretty big and convenient as Chapel's default int value (see #17643 for details).  It also adds a `.sizeAs()` routine for more explicitly getting the size as a specific integer type.  These changes have some ripple effects, such as using integers to store internal fields relating to an array's sizes (which is good because a multidimensional array with `idxType` of `int(8)` could've easily overflowed these values before) and changing `shape()` queries on domains and arrays to similarly return tuples of `int` rather than `idxType`.

The main pain of this PR is that in order to avoid pulling the rug out from users, we want to gently deprecate the old behavior.  At the same time, we want to give users the ability to get their code ready for the new world.  As a result, a `config param` has been added which allows one to opt into the new behavior now rather than waiting.  When this `config` is unset, and the user is using a non-`int` `idxType`, they will get a warning letting them know about the change.  To avoid this warning, they either need to opt into the `config` or switch to the `.sizeAs()` routine (or squash all warnings).

The impact of this warning is that all uses of `.size` on ranges, domains, and arrays within our module code needed to change to `sizeAs()` routines to ensure that they would not get the warning.  And this is a bit nerve-wracking because it's hard to feel confident that we got all such calls since the warnings will only be generated for code paths our tests cover; and because there are other `.size` queries on tuples, lists, etc. that did not need to change.

One specific case that is not currently handled is that the `.shape` routine does not have a user-facing `.shapeAs()` equivalent (primarily because it seems like silly overkill), yet library routines like LinearAlgebra use the `.shape()` query, so could generate errors if users call the routines on domains/arrays whose `idxType` is not `int`.  This seems sufficiently unlikely to me that I did not put any effort into it, but will check with @ben-albrecht and others to see whether they agree.

The other slightly challenging decision that had to be made in this transformation was what types module uses of `.sizeAs() should use, where the most obvious choices would be:
* `.sizeAs(intIdxType)`: matches previous behavior, so is no worse than what we had (though we know there were cases where this was insufficient in the past, such as #17264 and #7530).
* `.sizeAs(int)`: seems attractive in that it will be the new default once we remove the old behavior; yet is potentially not big enough for all `range(uint)` values (it's also potentially not big enough for `domain(uint)` and arrays over such domains, though this seems much less likely since there will typically be arrays of that size, and that's a pretty huge array).
* `.sizeAs(uint)`: has the benefit of always being big enough, but the downside of using `uint`s which can be annoying in Chapel

In practice, the heuristic I typically took was to try and use `.sizeAs(uint)` when I thought it was likely/probable that the size would conservatively need `max(uint)` values; `.sizeAs(int)` in other cases; and `size(idxType)` when I felt the old behavior was reasonable or just didn't want to worry about it for now.  I strongly suspect that if one computed extensively with non-`int` ranges, domains, and arrays, weaknesses in these choices could be found, though I'm not sure that things are much worse than they were before (where, already, using `uint` indices could often lead to problems), and have a slight hope that I've improved some things.

Not directly related to this change, but incorporated as a result of reviewer feedback, I also refactored the `.safeCast()` routine so that when checks are _on_, the amount of code inlined is much smaller by refactoring a lot of the bounds checking code into a helper routine.  This was motivated in this code by my reliance on a greater number of .safeCast() calls in the DR array implementation.

The changes involved in this are as follows:

Updated ChapelRange:
* added the `sizeReturnsInt` config to alternate between the old and new behaviors
* updated `.size` to work as described above
* added `range.sizeAs()`
* changed cases of `.size` to one of `size(idxType)` or `size(uint)` based on context
* fixed a case of bad recursion for the `.high` routine for single-value idxTypes by reasoning about `_low` and `_high` rather than `.size`

Updated ChapelArray:
* implemented similar `.size`/`.sizeAs` routines for domains and arrays
* similarly changed calls to `.size` into `.sizeAs()`
* changed .shape routines to have similar "warn" / "permit using `rank*int`" using the same `sizeReturnsInt` config
* added a `.chpl_shapeAs()` routine for internal uses of `shape()` that want to avoid the warning but wasn't convinced we'd want to support this long-term, so didn't expose it to the user

Updated storage of DefaultRectangular:
* changed block multipliers, origin, and factoredOffs from intIdxType to simply `int` (large, convenient, supports signed and unsigned)
* similarly, generally changed math used to compute offsets into an array's buffer from using `intIdxType` to using `int`
* changed uses of `.size` to `.sizeAs()`

For array implementation / range utility code:
* changed multipliers / address math to generally use `int` types rather than `idxType`
* changed fairly general purpose `.size` calls to `.sizeAs()` or to avoid warnings
* changed `.size:someIntType` patterns into `.sizeAs(someIntType)`

...in these files:
* modules/dists/BlockDist.chpl
* modules/dists/CyclicDist.chpl
* modules/dists/DSIUtil.chpl
* modules/dists/DimensionalDist2D.chpl
* modules/dists/ReplicatedDist.chpl
* modules/dists/dims/BlockCycDim.chpl
* modules/dists/dims/BlockDim.chpl
* modules/dists/dims/ReplicatedDim.chpl
* modules/internal/ArrayViewReindex.chpl
* modules/packages/RangeChunk.chpl
* modules/standard/DynamicIters.chpl

In Random.chpl:
* changed `.size` calls into `.sizeAs()` calls

Change the compiler's lowering of coforall loops to use a helper routine `chpl_coforallSize` to compute the size of a loop rather than relying on a direct call to `.size()` to avoid the warning that we generate for .size() calls today:
* compiler/AST/build.cpp
* modules/internal/ChapelBase.chpl

Updated test to fix creation of expected result when size == 0 (previously had been passing due to .size wrapping around):
* test/distributions/vass/densify-1-self.chpl

Added compopts file to switch to new behavior early:
* test/distributions/robust/arithmetic/trivial/test_dot_numIndices.compopts
* test/types/range/diten/testRangeLength.compopts
* test/types/range/elliot/rangeLength/intRangeLength.compopts
* test/types/range/elliot/rangeLength/rangeLengthOverflow.compopts
* test/types/range/elliot/rangeLength/uintRangeLength.compopts
* test/types/range/userAPI/COMPOPTS

Retired future that now works and created a variant that tests generated error messages:
* test/types/range/elliot/rangeLength/rangeLengthOverflow.bad
* test/types/range/elliot/rangeLength/rangeLengthOverflow2.*

Updated test to keep it working better than before:
* test/types/range/elliot/rangeLength/uintRangeLength.*

Added test to lock in that .size generates `int`:
* test/types/range/sizeType.chpl

Updated range API test to test .sizeAs() routine:
* test/types/range/userAPI/*

Updated array API test to lock in new warnings:
* test/arrays/userAPI/unsignedOps.good

Added a test locking in the resolution of #17264:
* test/arrays/bugs/sizeExceedsIntIdxType.*

Added a test to remind me to remove the old behavior:
* test/deprecated/rangeSize.chpl

Updated `.size` calls for non-`int` idxType tests to use `.sizeAs()` instead:
* test/library/packages/Sort/correctness/shellSortExtremeInds.chpl

Updated `.size` calls for non-`int` idxType tests to use `.sizeAs()` instead, and also addressed a reliance on signed int overflow:
* test/library/packages/Sort/correctness/sortAllIdxTypes.chpl

Refactor `.safeCast()` to reduce inlined size when checks are _on_:
* modules/standard/Types.chpl
* test/modules/sungeun/init/printModuleInitOrder*.good
* test/optimizations/deadCodeElimination/elliot/countDeadModules*.good
 
Resolves #7530
Resolves #17264 

TODO before merging this PR:
- [x] scan for `.sizeAs(int) == 0` or `.sizeAs(uint) cases in files prior to ChapelRange.chpl
- [x] gently change domain.shape / array.shape to return int or idxType
- [x] extend domain/array .size to work similarly (and dsiNumIndices)
- [ ] should we make uses of `.shape` in modules (e.g., LinearAlgebra.chpl) more bulletproof, or assume nobody will notice/mind?
- [x] update documentation for routines that changed
- [x] check that docs for config are rendered properly
- [x] ~consider making blk multipliers into uints?~ (tried this, decided it was more difficult than I was ready for in this PR)
- [x] ~performance playground?~ (decided, with the perf team, that since we don't really have many uses of arrays with non-int idxType in benchmarks that this was overkill)
- [x] run Cyclic distribution robustness tests with GASNet enabled
- [x] Do I have tests locking in new error messages?
- [x] Add test for #12764

As a clean-up after these changes have been out for a release, we can:
- [ ] convert `.sizeAs(int)` calls to `.size` for simplicity

### CODE REVIEW

**Comment by bradcray** [comment]
@damianmoz / @ben-albrecht : I have questions about this effort for you guys

@damianmoz: The question for you is whether you're using LinearAlgebra routines on any arrays or domains whose indices are not default integers.  This would happen if you declared a domain using another idxType like `domain(int(8))` or `domain(uint)` or if you defined a domain or array using bounds that were not integers, e.g.:  `var d = {1..myUint, 1..myUint};`

@ben-albrecht : The question for you is the following:  The `.shape()` query on domains and arrays currently issues a warning if the idxType is anything other than `int`.  For a few internal cases, I introduced a short-term `.chpl_shapeAs()` routine that permits one to bypass this warning, similar to the `.sizeAs()` query for ranges, domains, and arrays.  But I haven't gone to the trouble of updating all of our modules' uses of `.shape` (where LinearAlgebra is one of the main contenders) because my strong suspicion is that use cases in the field aren't likely to use non-`int` indices, so it would be a bunch of work that isn't really necessary and would need to be undone later.  Thoughts on that?

**Comment by damianmoz** [comment]
For now, I have stopped using LinearAlgebra. Two others colleagues (clients, not employees) who were using it and hence Chapel, have dropped Chapel entirely as a result of the 0-base mess.

I do not use default integers (or anything) in my own code. But I am lazy are they sometimes creep into my unit tests, Naughty, very naughty Damian.

I generally use 32 bit integers for array bounds. Occasionally 16 bits, or unsigned 32 bits, and very occasionally 8 bit. If I use 64 bits integers for an array bound, it is normally my screw-up. I love the fact that Chapel lets me do
```
const n = 20:int(32);
var a : [1..n, 1..n] real(64)
```
and has the indices as spanning int(32). If I had to do 1:int(32) in the above, I would not be a happy bunny.

File: test/library/packages/Sort/correctness/sortAllIdxTypes.chpl
Comment (mppf) [inline test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Noting that PR #17750 will also change this test. The change here in your PR still has signed integer overflow.

**Review by mppf — APPROVED** [review]

File: test/library/packages/Sort/correctness/sortAllIdxTypes.chpl
Comment (bradcray) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Good point.  We could change this to `.sizeAs(int)` instead?  (I updated it in this way simply out of laziness—trying to preserve the previous behavior as much as possible without thinking about it much).

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Since there will be a conflict with PR #17750 either way I think you should just leave it as-is and try to merge this before my PR :)

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Oh, and maybe it wasn't obvious, but the change I made in that PR takes away the `.size` calls altogether.

Comment (bradcray) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: I admittedly haven't gotten my head around #17750 yet.  If we used `sizeAs(int)` here, would the integer overflow still occur?  If so, can you spell out for me a bit better what it is?  (My assumption from the reproducer on https://github.com/chapel-lang/chapel/issues/6160#issuecomment-842424446 was that the problem was that `.size` was returning `max(int(8))` before and then adding to / subtracting from it; but if it returned an `int(64)` — which is the same type stored in the array — wouldn't the overflow go away?

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Right, if we used a `size` returning an `int` then there wouldn't be overflow. The reason I adjusted the test in a different manner from that is that there might still be truncation when converting back to e.g. `int(8)` - meaning the test would still rely on 2's complement etc. when in reality I think we just wanted the indices in a non-sorted order. I though the test would be nicer if it was independent of such issues so I adjusted it to always use the reverse range iteration to get the starting array.

Comment (bradcray) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: > when converting back to e.g. int(8)

When would that conversion back occur?  For this statement, `int - int(8)` would use the `-(int, int)` case and then `C[i]` stores `int`.  But is there a conversion back later that I'm missing?

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: Oh, you are right. I was imagining that we were storing to an `int(8)` element but we're not. However it's  still the case that the input array has the numbers in a harder to understand order based on 2's complement.

Comment (bradcray) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: > However it's arguably nicer for the test to have the values to sort in the original input range.

Oh, that's a good point.  So to make it safe in that regard (with a minimal modification to what's here), I think we'd need to do something like:

```chapel
    C[i] = (C.sizeAs(uint(s))-i:uint(s)):int(s);
```

does that seem right?  (rationale: `uint(s)` math is guaranteed to wrap around appropriately; then the cast to `int(s)` gets it back to the correct index set; then an implicit coercion to `int` takes place in the assignment.

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl:28]: I think that would avoid the signed overflow issue, yes; at least with my understanding of #6160 which I still consider open.

**Comment by bradcray** [comment]
@ronawho: Are you interested in also reviewing this PR at all, from the perspective of the impact on range overflow issues (which you've been invested in in the past) or potential performance in the DR code paths? (where I mentioned on chat that the increased use of `.safeCast()`s _could_ have a negative impact).

**Comment by ronawho** [comment]
> @ronawho: Are you interested in also reviewing this PR at all, from the perspective of the impact on range overflow issues (which you've been invested in in the past) or potential performance in the DR code paths? (where I mentioned on chat that the increased use of .safeCast()s could have a negative impact).

Yup, happy to review. Probably won't have time until Tuesday afternoon though. Is `safeCast` just a normal case with `--fast` or are those always on checks?

**Comment by bradcray** [comment]
> Is safeCast just a normal case with --fast or are those always on checks?

I think the former (modulo that the code path taken is different?), but TBH, I didn't check.  I wanted to write it the way I thought it should be right with fingers crossed that it wouldn't have performance impacts; but this was the reason I was asking whether we'd want to run it through the perf playground before merging (where I was misremembering, thinking that the only DR change had been the changes in field types).

File: test/library/packages/Sort/correctness/sortAllIdxTypes.chpl
Comment (bradcray) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl]: @mppf: I ended up making this change since (a) I feel responsible for having introduced a bug with this test and (b) it felt thematically related to the `.sizeAs()` routine (where I'd wrestled with phrasing this line a few different ways in this PR prior to the version you reviewed).  But I hadn't realized that you'd reviewed and approved this PR prior to doing so, so wanted to mention it here so that you weren't caught off-guard.

Comment (mppf) [reply test/library/packages/Sort/correctness/sortAllIdxTypes.chpl]: For this Pr, I'm fine with any version that passes testing. Let's talk about what we want to end up with under #17750.

**Comment by ben-albrecht** [comment]
> One specific case that is not currently handled is that the .shape routine does not have a user-facing .shapeAs() equivalent (primarily because it seems like silly overkill), yet library routines like LinearAlgebra use the .shape() query, so could generate errors if users call the routines on domains/arrays whose idxType is not int. This seems sufficiently unlikely to me that I did not put any effort into it, but will check with @ben-albrecht and others to see whether they agree.

I agree that it is unlikely to cause issues and is not worth the effort until a compelling use-case shows up (if ever).

**Comment by damianmoz** [comment]
I do not like that I cannot return the size as the idxType because I have had problems with indexing before where the compiler has complained about type problems in some element of a constructed range. I can see that the proposal solves some problems but it might create others. While it is probably a whole other issue, I would like to be able to turn Chapel's default type coercions off altogether at the command line in some cases as it might help find errors.

If the size is going to loose the native type of the range, you at least need to be able to return the size as its native type without having to remember the type to feed into the `sizeAs()` method. Does that mean you need another method called say `sizeAsNative`?

Just my 2c.

Why do I have problems finding this (I assume) PR.  I typed #17628 (and nothing else) in the search box and it did not find it directly.

**Comment by bradcray** [comment]
Hi @damianmoz —

> I love the fact that Chapel lets me do...

Happily, nothing about these patterns is changing here.

> I do not like that I cannot return the size as the idxType...

Under this PR, you can still get the size as an `idxType` value when desired, though it's admittedly more work:

* `r.size:idxType`
* `r.sizeAs(r.idxType)`
* `r.size.safeCast(r.idxType)`

We could consider adding a `.sizeAsNative`, `.sizeAsIdxType` style query as a convenience but I'd like to understand the motivating cases a bit better before adding them to the language.  Note that a user could always implement their own utility functions like this:

```chapel
proc range.sizeAsNative
  return this.sizeAs(this.idxType);
```

> ...because I have had problems with indexing before where the compiler has complained about type problems in some element of a constructed range.

I'm not snapping to what these patterns might be and how they might be helped or hurt by this PR.  If you can provide more details, I'd be interested in understanding better.

Note that `r.low` and `r.high` will still return their values as `idxType` like always, where I'm hypothesizing that most indexing expressions which are based on queries from a range will typically use those values rather than `.size` to compute an index.  For example, `A[A.domain.dim(0).high]` is a better pattern in Chapel than `A[A.domain.dim(0).size]` because it's index-neutral and guaranteed to be in-bounds.

> I can see that the proposal solves some problems but it might create others.

From the changes I've made to our code base, it seems clear to me that the net positives of this change outweigh the inconveniences, where one of my favorites is that I can now query the size of the following range:

```chapel
var n: uint(8) = 126
var r = (-n..n);
```

(where, before, it [resulted in a negative value](https://tio.run/#chapel)).

> While it is probably a whole other issue, I would like to be able to turn Chapel's default type coercions off altogether at the command line in some cases as it might help find errors.

Yeah, let's move that to its own issue if you're interested in it.

> Why do I have problems finding this (I assume) PR. I typed #17628 (and nothing else) in the search box and it did not find it directly.

I don't personally have any great tips for searching for PRs or issues by number in github.  Most often, I hack the URL directly if I don't have an existing link from email or the like to click on.

**Comment by damianmoz** [comment]
This time I went fishing with is:pr and #17628 and it worked. How do you know something coming from github via email is a PR unless somebody helpfully writes PR in the text?

Anyway, back to the topic.

To aid your snapping, see #17756. I cannot remember the exact details but their was an unsigned integer and a signed integer in the mix. And the error message was less than helpful but I was less than smart in what I had done. And I was doing it in a hurry to test something else. But other than that, I cannot remember much.

All your suggestions to return size demand that I use `r.idxType`

* `r.size:r.idxType`
* `r.sizeAs(r.idxType)`
* `r.size.safeCast(r.idxType)`

which is counterproductive and will break some of my existing code, most of which uses int(32).

I do not disagree that using **int** solves a lot of problems. I do not understand how (-126..126) is not 252 and a valid **int**(8).

But what happens when people start using **int**(128) when that exists?

Also, I note that your example returns correct results with 1.22.0. Or am I doing something really wrong?
```chapel
const n = 126:uint(8);
// or const n : uint(8) = 126 (which seems odd to me because it demands the compiler coerce/cast something)
// note that const t = 1.0:real(32) is legitimate but const t : real(32) = 1.0 will fail to compile without a cast of 1.0
var r = (-n .. n);

writeln(r.low);
writeln(r.high);
writeln(r.size);
```
It even works with **127** instead of **126**.

**Comment by bradcray** [comment]
> I do not understand how (-126..126) is not 252 and a valid int(8).

Signed `int(8)`s can only store values in the range -128 through 127, so 252 isn't a valid value it can store.  The situation is exacerbated with multidimensional domains/arrays, where it's much easier for a 3D `int(w)` domain or array to have more than `max(int(w))` elements.

> I note that your example returns correct results with 1.22.0. Or am I doing something really wrong?

Not _really_ wrong, but you declared `n` to be a `uint(8)` rather than an `int(8)` as in my example:

> 
> ```chapel
> > const n = 126:uint(8);
> ```

Changing that to `int(8)` gets the bad behavior I was advertising.

Why does it work with a `uint(8)` value of `n`?  Because unary `-` on `uint(8)` promotes the expression to `int(16)` (since `uint(8)` can't store negative numbers and `int(8)` isn't guaranteed to be able to safely store the result).  So you're essentially creating a range with `int(16)` idxType, which is big enough to store the size of 252 (yet, analogously, not big enough to store a range that spanned from `-(max(int(16))-1)..max(int(16))-1`).

> which is counterproductive and will break some of my existing code, most of which uses int(32).

I agree that the workarounds are a hassle / more typing, but I'm not understanding how using them would break your existing code?  Or are you saying that _not_ using them would break your code once this PR was merged?  (which I won't argue with, though I'd be curious to see what kinds of specific idioms in your code would be sensitive to this change).

Again, my recommendation is that if that's the query you commonly want, to define your own `.sizeAsXYZ` query on ranges that does what you want, as sketched out in my previous comment.

> To aid your snapping, see #17756.

Well, I'm trying to fix that issue as well.  That said, I don't expect to make all cases that mix types "just work".  Specifically, I don't believe there's a "good"/"right" answer for what adding an `int(64)` and a `uint(64)` should do, so believe that should continue to be an error.

> But what happens when people start using int(128) when that exists?

Personally, I don't expect that we will ever change the default `int` width from 64 bits, because I think it would be too significant of a breaking change.

> How do you know something coming from github via email is a PR unless somebody helpfully writes PR in the text?

I'm not sure... I don't subscribe to most github notifications.

File: compiler/AST/build.cpp
Comment (ronawho) [inline compiler/AST/build.cpp:1173]: I think this could have been changed to use `sizeas(int)`. I don't think there's any need to preserve the index type here. I'm fine leaving it as-is, but I was confused why this change was needed at first.

File: modules/internal/ChapelArray.chpl
Comment (ronawho) [inline modules/internal/ChapelArray.chpl:1839]: Not that it really matters since it's just to help us categorize halt, but I wonder if this is more appropriate as a castCheckHalt or maybe it's called safeCastCheckHalt. 

That also makes me wonder if this should only be enabled with boundsChecks or safeCastChecks instead of always on.

Comment (ronawho) [reply modules/internal/ChapelArray.chpl:1839]: Do we need to do any checking for negative numbers here? Assuming no since size should always be >=0, but I'm wondering if some user defined distribution could use negative size as a sentinel for something.

File: modules/internal/ChapelArray.chpl
Comment (ronawho) [inline modules/internal/ChapelArray.chpl:1815]: I wonder if `size` and `sizeas` should be inlined. May be a bad idea for the overflow check in sizeas, but we could throw that into a helper function

File: modules/internal/ChapelRange.chpl
Comment (ronawho) [inline modules/internal/ChapelRange.chpl:177]: Seems slightly odd for this to be in ChapelRange when it applies to ranges/domains/arrays. I wonder if chapelbase or something else makes more sense, but we probably don't chpldoc chapelbase.

File: modules/internal/ChapelArray.chpl
Comment (ronawho) [reply modules/internal/ChapelArray.chpl:1839]: In chapelrange this is guarded by `boundsChecking`, so I feel more confident we should use only do this check with runtime checks enabled.

**Comment by ronawho** [comment]
>> Is safeCast just a normal case with --fast or are those always on checks?

> I think the former (modulo that the code path taken is different?), but TBH, I didn't check. I wanted to write it the way I thought it should be right with fingers crossed that it wouldn't have performance impacts; but this was the reason I was asking whether we'd want to run it through the perf playground before merging (where I was misremembering, thinking that the only DR change had been the changes in field types).

> Edit: Taking a quick look, it appears that it compacts down to a normal cast in the event that checks are off. But it's a long routine, so I'm not 100% confident in that.

From an execution time performance perspective, I agree it is inlined and just a regular cast. However, since it is inlined and the routine is pretty big I wonder if this will have an impact on generated code size for compiles with checks on. I think we should move the checks into a helper function to avoid that. Something like:


```chpl
inline proc integral.safeCast(type T: integral) : T {
  if castChecking then this.safeCastChecks();
  return this:T;
}
```

File: compiler/AST/build.cpp
Comment (bradcray) [reply compiler/AST/build.cpp:1173]: I think you may be right; I did it this way less to get `sizeAs(idxType)` and more because it seemed simpler to move the code to a module and edit it there over time rather than having to build up AST expressions in the compiler.  It was also one of the only .size queries in the compiler, so this had the benefit of getting them all into module code.

File: modules/internal/ChapelArray.chpl
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:1839]: Good catch on the missing bounds-check test.

I was considering this a bounds check halt in the sense that the value we're trying to store into an integer is out of bounds for that integer type.  But I agree that it's in the grey area between bounds checking and safe casting.  I'm inclined to leave it alone if you're OK with that.

Interesting question about negative numbers.  If an implementation did use a negative size as a sentinel, is that something we should preserve and pass through or complain about here in the generic code?  I don't feel like I know what's right, so am inclined to leave that to the first person to run into a problem with having the dsi routine return a negative number to wrestle with.

File: modules/internal/ChapelArray.chpl
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:1815]: I'm going to argue "no" for this PR, based on the following:
* it wasn't prior to these changes when it was arguably simpler
* it's not clear to me that it will appear in time-critical code
* for it to be valuable, I suspect we'd also need to inline the dsiNum*() routines as well?
* I've become somewhat wary of premature optimization through inlining

If someone else did a study that found that inlining these was beneficial to compile-time or execution-time, I definitely wouldn't fight it, but it doesn't feel like it fits this PR's theme or is something we should take on without more evidence.

File: modules/internal/ChapelRange.chpl
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:177]: I think you're right that we don't (document ChapelBase), and think that we should eventually.

I don't think it's out of place here in the sense that arrays depend on domains and domains depend on ranges, so it feels at an appropriate point in the code hierarchy.  It's a bit odder in terms of the documentation hierarchy, but since it's a short-term feature by design, and one that hopefully most users won't ever have to use (because they use `int` idxTypes in their code almost all of the time... fingers crossed), I don't feel too sheepish about it being here.  And I think I got the docs in the other modules to correctly cross-link to the documentation in ChapelRange such that people will find their way here when need be (?).

**Comment by bradcray** [comment]
> I think we should move the checks into a helper function to avoid that.

Good thought.  This should help with existing code as well (I count 270 safeCast() calls in modules code).  I just pushed this.

File: modules/internal/ChapelArray.chpl
Comment (ronawho) [reply modules/internal/ChapelArray.chpl:1839]: I'm fine leaving it as a bounds check halt. 

I'm also ok leaving the question of negative numbers for later

File: modules/internal/ChapelRange.chpl
Comment (ronawho) [reply modules/internal/ChapelRange.chpl:177]: Sounds good.

File: modules/internal/ChapelArray.chpl
Comment (ronawho) [reply modules/internal/ChapelArray.chpl:1815]: Sounds good, and I agree with not wanting to inline prematurely. The case I was wondering about is whether we call .size when indexing into arrays and if there's any benefit there. Or really I was thinking about it for block dist dsiAccess, but regardless I agree with saving this for a different day and checking that there would be a perf benefit first.

**Review by ronawho — APPROVED** [review]

**Comment by bradcray** [comment]
I'm going to re-test once before merging, and then try to merge tonight to get all my changes in on the same night (and an otherwise quiet-ish day).

**Comment by bradcray** [comment]
Hmm, got some minor, last-minute errors due to updating to master, so I'm going to give this another look tomorrow to make sure there are no surprises.

**Pull request merged**
